### PR TITLE
Add presentation options for preferred navigation bar hidden state

### PR DESCRIPTION
**Pull request by mayur8657**
### What
To be used when the presenting view controller is a navigation controller and we want to hide its navigation bar. We added 2 separate options so that we act on them only if they are explicitly provided to not override any existing customizations made inside presentables.

The options are internal so that they can be used only through the static function which implies that you should have only one of them at a time (but can be workarounded of course).

### Why
We landed in a situation where the controllers pushed in navigation stack, have a different preferences of UINavigationBar visibility. We believe PresentationOption is the best way to specify, if the Presentation prefers the UINavigationBar hidden when it is presented.

### CODE REVIEW

File: Presentation/PresentationOptions.swift [hunk 1]
```diff
@@ -99,6 +103,14 @@ public extension UIViewController {
     }
 }
 
+internal extension PresentationOptions {
+    /// Will hide the navigation bar
+    static let prefersNavigationBarHidden = PresentationOptions()
+
+    /// Will show the navigation bar
+    static let prefersNavigationBarShown = PresentationOptions()
```
Comment (fennecos) [inline Presentation/PresentationOptions.swift]: Should not this be the default? do we need this option?
It makes it possible to call `present(vc, options: [. prefersNavigationBarHidden, . prefersNavigationBarShown])`
What should be the behavior in this case?

Comment (mayur8657) [reply Presentation/PresentationOptions.swift]: These options are in the internal extension of PresentationOptions. So you cannot specify them directly together like this. you have to use a static function instead, which is kind of intuitive that you should only use one. But if you workaround it like 
`present(vc, options: [PresentationOptions. prefersNavigationBarHidden(true), PresentationOptions. prefersNavigationBarHidden(true)])`

It will consider the hidden behaviour as priority. We are considering logging the ambiguity here.

And to answer, "Do we need this option?" I would say, this is our attempt to push presentations into navigation stack, where some prefer it hidden, and some don't.

We would love to know if there is any other approach we should take to achieve this.

Comment (mayur8657) [reply Presentation/PresentationOptions.swift]: Also this is supposed to add only preference for the Presentation. If unspecified, we would want the existing navigationController specifications to be kept. This is not to introduce a breaking change where the NavigationBar appearance is setup after the instantiation of the UINavigationController

Comment (fennecos) [reply Presentation/PresentationOptions.swift]: Ok I understand. Thanks for the clarification.

Is this only applied when pushing to a navigationController? Should it also work when modally presented?

File: Presentation/UINavigationController+Presenting.swift [hunk 2]
```diff
@@ -26,6 +26,13 @@ public extension PresentationOptions {
 extension UINavigationController: PresentingViewController {
     public func present(_ vc: UIViewController, options: PresentationOptions) -> PresentingViewController.Result {
         let dismissFuture = vc.installDismissButton().future.map { throw PresentError.dismissed }
+
+        if options.contains(.prefersNavigationBarHidden) {
+            self.isNavigationBarHidden = true
+        } else if options.contains(.prefersNavigationBarShown) {
+            self.isNavigationBarHidden = false
+        }
+
```
Comment (mansbernhardt) [inline Presentation/UINavigationController+Presenting.swift]: So what happens if we push another controller with none of the above options set? Will the navigation bar be hidden or not in that case or rather be dependent on the previous pushed view controller? Are the "non of options above" a way to allow users setting `isNavigationBarHidden` explicitly form the out-side. What if we say that setting `isNavigationBarHidden` from the out-side is not supported, would it make sense to only have `.prefersNavigationBarHidden` and change the above to:

```swift
isNavigationBarHidden = options.contains(.prefersNavigationBarHidden)
```

Comment (mayur8657) [reply Presentation/UINavigationController+Presenting.swift]: The state of `isNavigationBarHidden` remains unchanged until the preference option is specified in the `PresentationOptions` of next presentation.

If we think of UIKit way of doing it, we can always instantiate `UINavigationController` and hide the `UINavigationBar` on the `UINavigationController` object. And when a `UIViewController` object wants to hide it, we simply access `UIViewController` objects `navigationController` property. But that is usually done in the life cycle methods (after the ViewController is actually in the navigation stack). But as we do not play around with the UIViewController view's life-cycle methods, we felt presentation option is the way to go.

`isNavigationBarHidden = options.contains(.prefersNavigationBarHidden)`

This is the way we tried it before. But we realised, that it will be breaking the scenarios, pre-existent in the app. As we will have to specify the option everywhere, based on the preference. Hence we check for explicit specification of the controllers preference.

File: Presentation/UINavigationController+Presenting.swift [hunk 3]
```diff
@@ -26,6 +26,13 @@ public extension PresentationOptions {
 extension UINavigationController: PresentingViewController {
     public func present(_ vc: UIViewController, options: PresentationOptions) -> PresentingViewController.Result {
         let dismissFuture = vc.installDismissButton().future.map { throw PresentError.dismissed }
+
+        if options.contains(.prefersNavigationBarHidden) {
+            self.isNavigationBarHidden = true
+        } else if options.contains(.prefersNavigationBarShown) {
+            self.isNavigationBarHidden = false
```
[the remark this answers is not in the document]
Comment (mayur8657) [reply Presentation/UINavigationController+Presenting.swift]: Nice catch!

[the remark this answers is not in the document]
Comment (mayur8657) [reply Presentation/UINavigationController+Presenting.swift]: This has been fixed in the latest commit. With a tiny bit of glitch of late disappearance of navigationBar.

File: Presentation/PresentationOptions.swift (hunk 1 above)
Comment (mayur8657) [reply Presentation/PresentationOptions.swift]: It works even if modally presented.

File: Presentation/UINavigationController+Presenting.swift (hunk 2 above)
Comment (mansbernhardt) [reply Presentation/UINavigationController+Presenting.swift]: So are we currently restoring anything after a view controller has been displayed using these new options? I wonder if we should keep a state in the `PushPoper` helper of a view controllers's preferred navigation bar visibility and use that to restore when it becomes visible again such as after a pop?

Comment (mansbernhardt) [reply Presentation/UINavigationController+Presenting.swift]: I still would like to investigate if we can just settle on one `.prefersNavigationBarHidden` options but still make it work together with external changes. Not sure when an externally update is typically set. If it is set in materialize, we typically don't have a navigation controller available until perhaps we get the install callback. So if we store the preferred state based on options in PushPoper when being pushed, and then if the navigation bar visibility is externally changed while being presented and the view controller later is moved out of view (another controller is pushed), we can perhaps updated its PushPoper state if necessary. 

Could you elaborate the problems you currently hare having with mixing the options and external update?

File: Presentation/PresentationOptions.swift (hunk 1 above)
Comment (fennecos) [reply Presentation/PresentationOptions.swift]: so if we do:
```swift
present(vc, style: .modal, options: [.embedInNavigationController, .prefersNavigationBarHidden(true)])
```
I think that the navigation bar will be visible, since it's not calling present of UINavigationController

File: Presentation/UINavigationController+Presenting.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (fennecos) [reply Presentation/UINavigationController+Presenting.swift]: I see only one commit on this PR. not sure where the fix is. Could you point me to it?

[the remark this answers is not in the document]
Comment (mayur8657) [reply Presentation/UINavigationController+Presenting.swift]: @fennecos Can you see it now?

File: Presentation/PresentationOptions.swift (hunk 1 above)
Comment (mayur8657) [reply Presentation/PresentationOptions.swift]: Thanks! FIxed it now.

File: Presentation/UINavigationController+Presenting.swift (hunk 2 above)
Comment (mayur8657) [reply Presentation/UINavigationController+Presenting.swift]: We have added an example in the example project. That is typically the scenario we are trying to support.

ViewController1(Navbar visible) -> ViewController2(Navbar hidden) -> ViewController3(Navbar visible)

File: Presentation/UIViewController+Presentation.swift [hunk 4]
```diff
@@ -26,6 +26,7 @@ public extension UIViewController {
             }
         }
 
+        let previous = (self as? UINavigationController)?.isNavigationBarHidden
```
Comment (fennecos) [inline Presentation/UIViewController+Presentation.swift]: This working only if `present` is called from a `UINavigationController` but not from a `UIViewController` contained in a `UINavigationController`.
Let's say we have 
```swift
let nc = UIViewController()
let vc1 = UIViewController()
let vc2 = UIViewController()
nc.present(vc1)
vc1.present(vc2) // here vc2 will be pushed into nc, even though present is called on vc1
```

File: Presentation/UINavigationController+Presenting.swift [hunk 5]
```diff
@@ -26,6 +26,15 @@ public extension PresentationOptions {
 extension UINavigationController: PresentingViewController {
     public func present(_ vc: UIViewController, options: PresentationOptions) -> PresentingViewController.Result {
         let dismissFuture = vc.installDismissButton().future.map { throw PresentError.dismissed }
+
+        let current = self.isNavigationBarHidden
+        self.isNavigationBarHidden = options.contains(.prefersNavigationBarHidden)
+
+        let bag = DisposeBag()
+        bag += self.popViewControllerSignal.onValue { _ in
+            self.setNavigationBarHidden(true, animated: true)
```
[the remark this answers is not in the document]
Comment (mayur8657) [reply Presentation/UINavigationController+Presenting.swift]: I think this is committed by mistake while experimenting a bit. The current state should be passed on here.

File: Presentation/UIViewController+Presentation.swift (hunk 4 above)
Comment (mayur8657) [reply Presentation/UIViewController+Presentation.swift]: This is managed in UINavigationController+Presenting

File: Examples/StylesAndOptions/Example/NavigationExample.swift [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+struct TestNavigationBarHiding {
+
+}
+
+extension TestNavigationBarHiding: Presentable {
+    func materialize() -> (UIViewController, Disposable) {
+        let nc = UINavigationController()
+
+        let bag = DisposeBag()
+        bag += nc.present(Presentation(NavigationExample())).onValue {
```
[the remark this answers is not in the document]
Comment (mayur8657) [reply Examples/StylesAndOptions/Example/NavigationExample.swift]: flatMapLatest will cause problems as it will signal error when we go back. We need `flatMapIgnoringErrors` here. @iZNunev has already worked on it. But it needs to be merged in Flow.

**Comment by mayur8657** [comment]
@fennecos @mansbernhardt I made major changes in my approach. Please have a look now. This PR is currently blocking one of the tasks that needs to be part of next release.

File: Examples/StylesAndOptions/Example.xcodeproj/project.pbxproj [hunk 7]
```diff
@@ -363,6 +367,7 @@
 			buildSettings = {
 				ASSETCATALOG_COMPILER_APPICON_NAME = AppIcon;
 				CODE_SIGN_STYLE = Automatic;
+				DEVELOPMENT_TEAM = LBDSKKR6P2;
```
Comment (mansbernhardt) [inline Examples/StylesAndOptions/Example.xcodeproj/project.pbxproj]: Is this change intentional?

File: Presentation/PresentationOptions.swift [hunk 8]
```diff
@@ -46,6 +46,10 @@ public extension PresentationOptions {
 
     /// Default options used unless any options are explicity passed when presented: `[embedInNavigationController]`
     static let defaults: PresentationOptions = [embedInNavigationController]
+
+    static func prefersNavigationBarHidden(_ hidden: Bool) -> PresentationOptions {
+        return hidden ? navigationBarHidden : navigationBarShown
+    }
```
Comment (mansbernhardt) [inline Presentation/PresentationOptions.swift]: We discussed over slack if it was possible to just have one option `. prefersNavigationBarHidden`. I think that would be much more elegant. And as for backward compatibility, I suggested that the we keep the behavior we have now until a view controller is first pushed with `.prefersNavigationBarHidden`, where the user has stated its intention to use Presentation FW's handling navigation bar visibility.

So if the current push popper stack does not include any `.prefersNavigationBarHidden` disable navigation bar logic. But once a view controller is push with `.prefersNavigationBarHidden` enable logic.

So if the new top controller after processPushPopers has `.prefersNavigationBarHidden`, hide the navigation bar. And if the new top controller don't have `.prefersNavigationBarHidden` and any previous push popper (either just pop:ed or push:ed before), show the navigation bar. If none of these are true don't do anything with the navigation bar visibility (backward compatibility mode).

Do think this could work?

File: Presentation/UINavigationController+Presenting.swift [hunk 9]
```diff
@@ -78,6 +79,7 @@ public extension UINavigationController {
 
 private class PushPoper: NSObject {
     let vc: UIViewController
+    let vcPrefersNavBarHidden: Bool?
     let animated: Bool
     let disableCoalecing: Bool
```
Comment (mansbernhardt) [inline Presentation/UINavigationController+Presenting.swift]: Instead of store vcPrefersNavBarHidden and disableCoalecing we could probably just store options instead? (this does not apply to animated though)

File: Presentation/UINavigationController+Presenting.swift [hunk 10]
```diff
@@ -156,21 +162,31 @@ private extension UINavigationController {
...
+                if let navBarHidden = navBarHidden {
+                    self.setNavigationBarHidden(navBarHidden, animated: animated)
+                }
             }
         } else {
             setViewControllers(vcs, animated: animated)
+            if let navBarHidden = navBarHidden {
+                self.setNavigationBarHidden(navBarHidden, animated: animated)
+            }
         }
```
Comment (mansbernhardt) [inline Presentation/UINavigationController+Presenting.swift]: Not sure I follow all updates to `setNavigationBarHidden` above. Should it not just be enough to update based on the latest topViewController (vcs.last)?

File: Presentation/UINavigationController+Presenting.swift [hunk 11]
```diff
@@ -204,6 +220,12 @@ private extension UINavigationController {
             pushPoper.vc.navigationItem.popCallbacker.callAll(with: ())
             pushPoper.onComplete(.success)
         }
+        pushPoper.bag += willPopViewControllerSignal.filter { $0 == pushPoper.vc }.onFirstValue { vc in
+            guard self.viewControllers.count > 1 else { return }
+            if let previousPushPoper = self.popSignalPushPopers.first(where: { $0.value?.vc == self.viewControllers.last }).flatMap({ $0.value }), let navBarHidden = previousPushPoper.vcPrefersNavBarHidden {
+                self.setNavigationBarHidden(navBarHidden, animated: previousPushPoper.animated)
+            }
+        }
```
Comment (mansbernhardt) [inline Presentation/UINavigationController+Presenting.swift]: Why do we need this?

File: Presentation/PresentationOptions.swift (hunk 8 above)
Comment (mayur8657) [reply Presentation/PresentationOptions.swift]: @mansbernhardt  Yes. I remember we talked about this. And when I tried to imagine all the use cases. I don't think, only having `.prefersNavigationBarHidden` as option suits the scenario below:

`navigationController(nav bar is hidden) -> vc1 -> vc2  -> vc3 (wants NavBar) -> vc4 (no Navbar) -> vc5 (no navbar)`

You cannot specify any option to show the navigationBar for VC3 in presentation option, if we only have `.prefersNavigationBarHidden`.

Thus allowing to either specify the preference or to not specify anything that will keep existing state makes more sense to me.

Please let me know your thoughts on this @nataliq

File: Presentation/UINavigationController+Presenting.swift (hunk 9 above)
Comment (mayur8657) [reply Presentation/UINavigationController+Presenting.swift]: Yes. `animated` is not considered directly here. I was also tempted to pass on the options directly to PushPoper. But I see there are many options that have nothing to do with PushPopers. Holding them inside PushPopers seemed unnecessary to me so I decided to keep PushPoper independent of those. Do we have any strong reason to hold all the options in PushPoper?

File: Presentation/PresentationOptions.swift [hunk 12]
```diff
@@ -85,6 +89,11 @@ public extension UIViewController {
         let nc = customNavigationController(options)
         nc.transferDebugPresentationInfo(from: self)
         nc.viewControllers = [ self ]
+        if options.contains(.navigationBarShown) {
```
[the remark this answers is not in the document]
Comment (mayur8657) [reply Presentation/PresentationOptions.swift]: This is intentional. While presenting a Presentation with options like `.embedInNavigationController`, `.prefersNavigationBarHidden`
